File: src/main/java/alfheim/common/entity/boss/ai/AILightning.java
package alfheim.common.entity.boss.ai;

import java.util.HashSet;
import java.util.List;
import java.util.Set;

import alexsocol.asjlib.ASJUtilities;
import alexsocol.asjlib.math.Vector3;
import alfheim.common.entity.EntityLightningMark;
import alfheim.common.entity.boss.EntityFlugel;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.util.ChunkCoordinates;

public class AILightning extends AIBase {

	public AILightning(EntityFlugel flugel, AITask task) {
		super(flugel, task);
	}
	
	public Set<EntityPlayer> getRandomPlayers() {
		List<EntityPlayer> players = flugel.getPlayersAround();
		if (players.isEmpty()) return new HashSet<EntityPlayer>(0);
		int count = flugel.worldObj.rand.nextInt(players.size()) + 1;
		Set<EntityPlayer> set = new HashSet<EntityPlayer>(count);
		while (count > 0) {
			int player = flugel.worldObj.rand.nextInt(players.size());
			if (!set.contains(players.get(player))) {
				set.add(players.get(player));
				--count;
			}
		}
		return set;
	}

	@Override
	public void startExecuting() {
		flugel.setAITaskTimer(20);
		for (EntityPlayer player : getRandomPlayers()) player.worldObj.spawnEntityInWorld(new EntityLightningMark(player.worldObj, player.posX, player.posY, player.posZ));
		if (flugel.isHardMode()) {
			ChunkCoordinates src = flugel.getSource();
			for (int i = 0; i < ASJUtilities.randInBounds(flugel.worldObj.rand, 5, 10); i++) {
				Vector3 vec3 = new Vector3(ASJUtilities.randInBounds(flugel.worldObj.rand, -EntityFlugel.RANGE, EntityFlugel.RANGE), ASJUtilities.randInBounds(flugel.worldObj.rand, -EntityFlugel.RANGE, EntityFlugel.RANGE), ASJUtilities.randInBounds(flugel.worldObj.rand, -EntityFlugel.RANGE, EntityFlugel.RANGE)).normalize().mul(EntityFlugel.RANGE);
				flugel.worldObj.spawnEntityInWorld(new EntityLightningMark(flugel.worldObj, src.posX + vec3.x, src.posY + vec3.y, src.posZ + vec3.z));
			}
		}
	}

	@Override
	public boolean continueExecuting() {
		return canContinue();
	}
}

File: src/main/java/alfheim/common/entity/boss/ai/AILightning.kt
package alfheim.common.entity.boss.ai;

import alexsocol.asjlib.ASJUtilities
import alfheim.common.entity.EntityLightningMark
import alfheim.common.entity.boss.EntityFlugel
import net.minecraft.entity.player.EntityPlayer
import vazkii.botania.common.core.helper.Vector3
import java.util.HashSet

class AILightning(flugel: EntityFlugel, task: AITask) : AIBase(flugel, task) {

	fun getRandomPlayers(): Set<EntityPlayer> {
		val players = flugel.getPlayersAround()
		if (players.isEmpty()) return HashSet<EntityPlayer>(0)
		var count = flugel.worldObj.rand.nextInt(players.size) + 1
		val set = HashSet<EntityPlayer>(count)
		while (count > 0) {
			val player = flugel.worldObj.rand.nextInt(players.size)
			if (!set.contains(players.get(player))) {
				set.add(players.get(player))
				--count
			}
		}
		return set
	}

	override fun startExecuting() {
		flugel.setAITaskTimer(20)
		for (player in getRandomPlayers()) player.worldObj.spawnEntityInWorld(EntityLightningMark(player.worldObj, player.posX, player.posY, player.posZ))
		if (flugel.isHardMode()) {
			val src = flugel.getSource()
			for (i in 0 until ASJUtilities.randInBounds(flugel.worldObj.rand, 5, 10)) {
				val vec3 = Vector3(ASJUtilities.randInBounds(flugel.worldObj.rand, -EntityFlugel.RANGE, EntityFlugel.RANGE).toDouble(), ASJUtilities.randInBounds(flugel.worldObj.rand, -EntityFlugel.RANGE, EntityFlugel.RANGE).toDouble(), ASJUtilities.randInBounds(flugel.worldObj.rand, -EntityFlugel.RANGE, EntityFlugel.RANGE).toDouble()).normalize().multiply(EntityFlugel.RANGE.toDouble())
				flugel.worldObj.spawnEntityInWorld(EntityLightningMark(flugel.worldObj, src.posX + vec3.x, src.posY + vec3.y, src.posZ + vec3.z))
			}
		}
	}

	override fun continueExecuting(): Boolean = canContinue()
}